src/interned.rs

```diff
@@ -208,42 +241,103 @@ where
+            }
+
             // We won any races so should intern the data
             Err(slot) => {
                 let zalsa = db.zalsa();
                 let table = zalsa.table();
+
+                // Record the durability of the current query on the interned value.
+                let durability = zalsa_local
+                    .active_query()
+                    .map(|(_, stamp)| stamp.durability)
+                    // If there is no active query this durability does not actually matter.
+                    .unwrap_or(Durability::MAX);
+
                 let id = zalsa_local.allocate(table, self.ingredient_index, |id| Value::<C> {
                     fields: unsafe { self.to_internal_data(assemble(id, key)) },
                     memos: Default::default(),
                     syncs: Default::default(),
+                    durability: AtomicU8::new(durability.as_u8()),
+                    // Record the revision we are interning in.
+                    first_interned_at: current_revision,
+                    last_interned_at: AtomicRevision::from(current_revision),
                 });
-                unsafe {
-                    lock.insert_in_slot(
-                        data_hash,
-                        slot,
-                        (
-                            table.get::<Value<C>>(id).fields.clone(),
-                            SharedValue::new(id),
-                        ),
-                    )
-                };
+
+                let value = (
+                    table.get::<Value<C>>(id).fields.clone(),
+                    SharedValue::new(id),
+                );
+
+                unsafe { lock.insert_in_slot(data_hash, slot, value) };
+
                 debug_assert_eq!(
                     data_hash,
                     self.key_map
                         .hasher()
                         .hash_one(table.get::<Value<C>>(id).fields.clone())
                 );
+
+                // Record a dependency on this value.
+                let index = self.database_key_index(id);
+                zalsa_local.report_tracked_read(
+                    index,
+                    durability,
+                    current_revision,
+                    InputAccumulatedValues::Empty,
+                );
+
                 id
             }
         }
     }
 
+    /// Returns the database key index for an interned value with the given id.
+    pub fn database_key_index(&self, id: Id) -> DatabaseKeyIndex {
+        DatabaseKeyIndex::new(self.ingredient_index, id)
+    }
+
     /// Lookup the data for an interned value based on its id.
     /// Rarely used since end-users generally carry a struct with a pointer directly
     /// to the interned item.
     pub fn data<'db>(&'db self, db: &'db dyn Database, id: Id) -> &'db C::Fields<'db> {
         let internal_data = db.zalsa().table().get::<Value<C>>(id);
+        let last_changed_revision = db.zalsa().last_changed_revision(internal_data.durability());
```

Comment:
```suggestion
        let zalsa = db.zalsa();
        let internal_data = zalsa.table().get::<Value<C>>(id);
        let last_changed_revision = zalsa.last_changed_revision(internal_data.durability());
```
to reduce the amount of dynamically dispatched calls